server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -523,30 +526,44 @@ protected String doReadString(final int charCount) throws IOException {
         return new String(charBuffer, 0, charCount);
     }
 
-    protected static int calculateByteLengthOfChars(final byte[] bytes, final int chars, final int start, final int limit)
-        throws IOException {
-        if (limit - start < chars) {
-            return -1; // not enough bytes to read chars
+    protected String tryReadStringFromBytes(final byte[] bytes, final int start, final int limit, final int chars) throws IOException {
+        final int end = start + chars;
+        if (limit < end) {
+            return null; // not enough bytes to read chars
         }
-        // calculate number of bytes matching the requested chars
-        int pos = start;
-        int remaining = chars;
-        while (pos < limit && remaining-- > 0) {
+        for (int pos = start; pos < end; pos++) {
+            if ((bytes[pos] & 0xf0) > 0x70) {
+                // not an ISO_8859_1 char, fall back to reading a UTF-8 string
```

Comment:
Nit: `ISO_8859_1` includes characters with the top bit set, we're just checking for `US-ASCII` here.